# Spark Funds Analysis : Assignment

## Business Objective

Spark funds is an asset management company and wants to make investments in a few companies. The CEO of Spark Funds wants to understand the global trends in investments so that she can take the investment decisions effectively. 
Our objective is to identify the best sectors, countries, and a suitable investment type for making investments. The overall strategy is to invest where others are investing, implying that the 'best' sectors and countries are the ones 'where most investors are investing'.

## Goals of the Analysis

Goals are divided into three sub-goals:

1. Investment type analysis: Comparing the typical investment amounts in the venture, seed, angel, private equity etc. so that Spark Funds can choose the type that is best suited for their strategy.

2. Country analysis: Identifying the countries which have been the most heavily invested in the past. These will be Spark Funds’ favourites as well.

3. Sector analysis: Understanding the distribution of investments across the eight main sectors.

## Data Description

Real time investment data from crunchbase.com is provided.
Three main data tables are provided which are as follows:
    1. companies.txt
    2. rounds2.csv
    3. mapping.csv
    
Further description about the data tables is below:
1. Company Details(companies.txt):
    A table with basic data of companies.
    Description of Companies Table <br>
    a. Permalink : Unique ID of company <br>
    b. name : Company name <br>
    c. homepage_url : Website URL <br>
    d. category_list : Category/categories to which a company belongs <br>
    e. status : Operational status <br>
    f. country_code : Country Code <br>
    g. state_code : State <br>
2. Funding round details(rounds2.csv):
   Includes the details related to the funding towards a company.
   Description of rounds2 Table <br>
   a. company_permalink : Unique ID of company <br>
   b. funding_round_permalink : Unique ID of funding round <br>
   c. funding_round_type : Type of funding – venture, angel, private equity etc. <br>
   d. funding_round_code : Round of venture funding (round A, B etc.) <br>
   e. funded_at : Date of funding <br>
   f. raised_amount_usd : Money raised in funding (USD) <br>
3. Sector Classification(mapping.csv):
   This file maps the numerous category names in the companies table (such 3D 
   printing,  aerospace, agriculture, etc.) to eight broad sector names. Its purpose is
   to simplify the analysis into eight sector buckets, rather than trying to analyse 
   hundreds of them.

## Data Cleaning

In [34]:
#Importing required libraries
import pandas as pd
import chardet
from IPython.display import display
import numpy as np
pd.set_option('display.float_format', '{:.2f}'.format)

In [35]:
#importing rounds2 file into a pandas dataframe
rounds2 = pd.read_csv("Data/rounds2.csv",encoding='Palmos')
#Filling missing values
rounds2['funding_round_code']=rounds2['funding_round_code'].fillna('')
rounds2['raised_amount_usd']=rounds2['raised_amount_usd'].fillna('0')
display(rounds2.head(5))

company_permalink  \
0                  /organization/-fame   
1               /ORGANIZATION/-QOUNTER   
2               /organization/-qounter   
3  /ORGANIZATION/-THE-ONE-OF-THEM-INC-   
4                /organization/0-6-com   

                           funding_round_permalink funding_round_type  \
0  /funding-round/9a01d05418af9f794eebff7ace91f638            venture   
1  /funding-round/22dacff496eb7acb2b901dec1dfe5633            venture   
2  /funding-round/b44fbb94153f6cdef13083530bb48030               seed   
3  /funding-round/650b8f704416801069bb178a1418776b            venture   
4  /funding-round/5727accaeaa57461bd22a9bdd945382d            venture   

  funding_round_code   funded_at raised_amount_usd  
0                  B  05-01-2015       10000000.00  
1                  A  14-10-2014                 0  
2                     01-03-2014         700000.00  
3                  B  30-01-2014        3406878.00  
4                  A  19-03-2008        2000000.00

> Number of unique companies in rounds2

In [36]:
#Make sure that there are no missing values in the columns company_permalink
print(rounds2["company_permalink"].isnull().values.any())
#Making the company_permalink column to have the first case in a word as Capital
rounds2["company_permalink"]=rounds2["company_permalink"].str.title()
print("Number of Unique companies in rounds2 is : ",rounds2['company_permalink'].nunique())

False
Number of Unique companies in rounds2 is :  66368


Here, Number of unique companies in rounds2 file found to be 66368. Further, I have also found that there are no missing values present in the company_permalink column.

In [37]:
#load companies.txt into a pandas dataframe
companies = pd.read_csv("Data/companies.txt", sep='\t',encoding='Palmos')
#Preprocessing : filling missing values
companies.category_list=companies.category_list.fillna('')
companies['country_code']=companies['country_code'].fillna('')
display(companies.head(5))

permalink                    name  \
0                  /Organization/-Fame                   #fame   
1               /Organization/-Qounter                :Qounter   
2  /Organization/-The-One-Of-Them-Inc-  (THE) ONE of THEM,Inc.   
3                /Organization/0-6-Com                 0-6.com   
4       /Organization/004-Technologies        004 Technologies   

                        homepage_url  \
0                 http://livfame.com   
1             http://www.qounter.com   
2                http://oneofthem.jp   
3                 http://www.0-6.com   
4  http://004gmbh.de/en/004-interact   

                                       category_list     status country_code  \
0                                              Media  operating          IND   
1  Application Platforms|Real Time|Social Network...  operating          USA   
2                                  Apps|Games|Mobile  operating                
3                                        Curated Web  operating          CHN   
4                                           Software  operating          USA   

  state_code                 region           city  founded_at  
0         16                 Mumbai         Mumbai         NaN  
1         DE             DE - Other  Delaware City  04-09-2014  
2        NaN                    NaN            NaN         NaN  
3         22                Beijing        Beijing  01-01-2007  
4         IL  Springfield, Illinois      Champaign  01-01-2010

> Number of unique companies in companies file

In [38]:
#Make sure that there are no missing values in the columns company_permalink
print(companies['permalink'].isnull().values.any())

print("The number of unique companies in companies file is : ",companies['permalink'].nunique())

companies=companies.sort_values('permalink')

False
The number of unique companies in companies file is :  66368


In case of companies file, there is no missing value in column permalink and in total there are 66368 unique companies.

> Columns that can be used as unique key for companies

In [39]:
print("Number of rows in companies df : ",companies.shape[0])

print("Number of rows unique values in permalink column : ",companies['permalink'].nunique())
#Check if unique values of permalink is equal to the total rows of companies df
if companies['permalink'].nunique()== companies.shape[0]:
    print("Column permalink can be used as a unique key for Identifying companies as it identifies each row correctly")

Number of rows in companies df :  66368
Number of rows unique values in permalink column :  66368
Column permalink can be used as a unique key for Identifying companies as it identifies each row correctly


Column permalink can be used as unique key for identifying companies as it is not null and also  identifies each row uniquely in the table.

> Check if are there any companies in rounds2 that are not in companies file

In [40]:
#Creating a function to check the difference between the companies
def check_company_difference(A,B):
    tmp = set(A)-set(B)
    if len(tmp)>0:
        if len(tmp)==1:
            print("There is {} company which are in rounds2 but not in companies file".format(len(tmp)))
        else:
            print("There are {} companies which are in rounds2 but not in companies file".format(len(tmp)))
        print("Below is the companies list :")
        print(tmp)
    else:
        print("There is no difference")

In [41]:
Companies_A = companies['permalink']
Companies_B = rounds2['company_permalink']
check_company_difference(Companies_B,Companies_A)

There is 1 company which are in rounds2 but not in companies file
Below is the companies list :
{'/Organization/Energystone-Games-Ç\x81Μçÿ³Æ¸¸Æˆ♥'}


On investing the file manually, I found that there is a company with the same name as above present in both the files. Thus, there might be a possibility of data processing issues which are causing the problem. Lets try to dig down more into the issue using the code:

In [42]:
Company_name_rounds2=rounds2[rounds2['company_permalink'].str.contains('Energystone-Games')]['company_permalink']
print(Company_name_rounds2)

31863    /Organization/Energystone-Games-ǁΜçÿ³Æ¸¸Æˆ♥
Name: company_permalink, dtype: object


In [43]:
Company_name_companies=companies[companies['permalink'].str.contains('Energystone-Games')]['permalink']
print(Company_name_companies)


18197    /Organization/Energystone-Games-ǁµçÿ³Æ¸¸Æˆ♥
Name: permalink, dtype: object


As I found that the due to encoding python is segmenting the two companies. To resolve this I will simply make the company name consistent across both the files by using the name of the company from the rounds2 file.

In [44]:

#As both the Names of the companies are same, I will make them consistent across both the files.
companies.iloc[18197,0]='/Organization/Energystone-Games-ǁΜçÿ³Æ¸¸Æˆ♥'

In [45]:
Companies_A = set(companies['permalink'])
Companies_B = set(rounds2['company_permalink'])
#Checking Again for the companies which are in rounds2 but not in companies file
check_company_difference(Companies_B,Companies_A)

There is no difference


Now, as we can see that there are no companies which are there in rounds2 and are not present in the companies file.

> Merging the two dataframes

In [46]:
master_frame = pd.merge(rounds2,companies,how='inner',left_on='company_permalink',right_on='permalink')
#Number of rows in master_frame
print("Number of rows in master_frame merged frame is {}".format(master_frame.shape[0]))

Number of rows in master_frame merged frame is 114949


Total number of rows after merging of the companies and rounds2 file is 114949.

## Funding Type Analysis

To find out the most representative value of the investment amount for each of the four funding types, we can group by the master_frame using the funding round type and compute the mean for each funding type. This mean value will help us to decide the best funding type based on the constraints of Spark Funds.

In [47]:
master_frame.raised_amount_usd=pd.to_numeric(master_frame.raised_amount_usd)
grouped=master_frame.groupby('funding_round_type')[['raised_amount_usd']].mean().reset_index()
grouped=grouped[grouped.funding_round_type.isin(['venture','seed','angel','private_equity'])]
print(grouped)


   funding_round_type  raised_amount_usd
0               angel          764564.35
8      private_equity        62111788.23
11               seed          556606.74
13            venture        10634054.44


The only funding_round_type that satisfies the Spark Funds constraints is venture as it has a funding round investment between 5 to 15 million USD.

We will filter out the master dataframe for only venture round now.

In [48]:
#Filtering the dataframe
master_frame=master_frame[master_frame.funding_round_type=='venture']
#checking whether the master_frame now contains only information for venture round
master_frame.funding_round_type.unique()

array(['venture'], dtype=object)

## Country Analysis

In [49]:
#removing countries with missing values
countries=master_frame[master_frame.country_code!='']
#Creating top9 dataframe
top9=countries.groupby('country_code')[['raised_amount_usd']].sum().sort_values(by=['raised_amount_usd'],ascending=False).head(9)
display(top9)

raised_amount_usd
country_code                   
USA             422510842796.00
CHN              39835418773.00
GBR              20245627416.00
IND              14391858718.00
CAN               9583332317.00
FRA               7259536732.00
ISR               6907514579.00
DEU               6346959822.00
JPN               3363676611.00

The above table represents the top9 countries which have received the highest total funding across all the sectors.
We can now use the information provided as a reference for determing the english speaking countries from the top9 dataframe.

In [50]:
#Create a list of english speaking countries in top9
eng_countries=['USA','GBR','IND','CAN']
top3=top9[top9.index.isin(eng_countries)]
top3=top3.sort_values('raised_amount_usd',ascending=False).head(3)
display(top3)

raised_amount_usd
country_code                   
USA             422510842796.00
GBR              20245627416.00
IND              14391858718.00

Now we will filter the information from the master_frame for these three countries.

In [51]:
master_frame=master_frame[master_frame.country_code.isin(list(top3.index))]
#Checking whether the master_frame now contains information for the top three english speaking countries
print(master_frame.country_code.unique())
display(master_frame.head(5))

['IND' 'USA' 'GBR']


company_permalink  \
0              /Organization/-Fame   
1           /Organization/-Qounter   
5   /Organization/004-Technologies   
10            /Organization/0Xdata   
11            /Organization/0Xdata   

                            funding_round_permalink funding_round_type  \
0   /funding-round/9a01d05418af9f794eebff7ace91f638            venture   
1   /funding-round/22dacff496eb7acb2b901dec1dfe5633            venture   
5   /funding-round/1278dd4e6a37fa4b7d7e06c21b3c1830            venture   
10  /funding-round/3bb2ee4a2d89251a10aaa735b1180e44            venture   
11  /funding-round/ae2a174c06517c2394aed45006322a7e            venture   

   funding_round_code   funded_at  raised_amount_usd  \
0                   B  05-01-2015        10000000.00   
1                   A  14-10-2014               0.00   
5                      24-07-2014               0.00   
10                  B  09-11-2015        20000000.00   
11                     03-01-2013         1700000.00   

                         permalink              name  \
0              /Organization/-Fame             #fame   
1           /Organization/-Qounter          :Qounter   
5   /Organization/004-Technologies  004 Technologies   
10            /Organization/0Xdata            H2O.ai   
11            /Organization/0Xdata            H2O.ai   

                         homepage_url  \
0                  http://livfame.com   
1              http://www.qounter.com   
5   http://004gmbh.de/en/004-interact   
10                     http://h2o.ai/   
11                     http://h2o.ai/   

                                        category_list     status country_code  \
0                                               Media  operating          IND   
1   Application Platforms|Real Time|Social Network...  operating          USA   
5                                            Software  operating          USA   
10                                          Analytics  operating          USA   
11                                          Analytics  operating          USA   

   state_code                 region           city  founded_at  
0          16                 Mumbai         Mumbai         NaN  
1          DE             DE - Other  Delaware City  04-09-2014  
5          IL  Springfield, Illinois      Champaign  01-01-2010  
10         CA            SF Bay Area  Mountain View  01-01-2011  
11         CA            SF Bay Area  Mountain View  01-01-2011

In [52]:
print(master_frame.shape[0])

41667


Now the master_frame only includes the information for the top3 english speaking countries with the highest funding and a fixed funding type as venture.

## Sector wise Analysis

### Sector Analysis 1

Now we need to identify the primary sector of each company from the master_frame followed by mapping the primary sector to main sectors provided in dataframe

In [53]:
# We will now split the column category list to find the primary sector for each company
master_frame['primary_sector']=master_frame.category_list.str.split('|',expand=True)[0]
display(master_frame[['company_permalink','primary_sector']].head(5))

company_permalink         primary_sector
0              /Organization/-Fame                  Media
1           /Organization/-Qounter  Application Platforms
5   /Organization/004-Technologies               Software
10            /Organization/0Xdata              Analytics
11            /Organization/0Xdata              Analytics

In [54]:
#load the mapping file
mapping=pd.read_csv("Data/mapping.csv")
#Filling na values in category_list column with zero
mapping.category_list = mapping.category_list.fillna('')
display(mapping.head(5))

category_list  Automotive & Sports  Blanks  Cleantech / Semiconductors  \
0                                   0       1                           0   
1             3D                    0       0                           0   
2    3D Printing                    0       0                           0   
3  3D Technology                    0       0                           0   
4     Accounting                    0       0                           0   

   Entertainment  Health  Manufacturing  News, Search and Messaging  Others  \
0              0       0              0                           0       0   
1              0       0              1                           0       0   
2              0       0              1                           0       0   
3              0       0              1                           0       0   
4              0       0              0                           0       0   

   Social, Finance, Analytics, Advertising  
0                                        0  
1                                        0  
2                                        0  
3                                        0  
4                                        1

While going through the file manually I found that in the category_list column of mappings file, "na" has been replaced with "0". Below are few instances of it.

In [55]:
mapping[(mapping.category_list.str.contains('0')==True)]

category_list  Automotive & Sports  Blanks  \
20            Alter0tive Medicine                    0       0   
22                       A0lytics                    0       0   
59              Big Data A0lytics                    0       0   
79              Business A0lytics                    0       0   
89                        Can0bis                    0       0   
91               Career Ma0gement                    0       0   
103                 Chi0 Internet                    0       0   
114               Cloud Ma0gement                    0       0   
145             Contact Ma0gement                    0       0   
198      Digital Rights Ma0gement                    0       0   
199                Digital Sig0ge                    0       0   
210            Document Ma0gement                    0       0   
223              Educatio0l Games                    0       0   
240              Energy Ma0gement                    0       0   
243                Enterprise 2.0                    0       0   
256               Event Ma0gement                    0       0   
273                        Fi0nce                    0       0   
274             Fi0nce Technology                    0       0   
275            Fi0ncial Exchanges                    0       0   
276             Fi0ncial Services                    0       0   
281               Fleet Ma0gement                    0       0   
310                     Gover0nce                    0       0   
353            Identity Ma0gement                    0       0   
370          Innovation Ma0gement                    0       0   
372  Intellectual Asset Ma0gement                    0       0   
384          Investment Ma0gement                    0       0   
389                  IT Ma0gement                    0       0   
391                     Jour0lism                    0       0   
394           Knowledge Ma0gement                    0       0   
400                Lead Ma0gement                    0       0   
437          Medical Professio0ls                    0       0   
448               Mobile A0lytics                    0       0   
471                 0notechnology                    0       0   
472    0tural Language Processing                    0       0   
473              0tural Resources                    0       0   
474                     0vigation                    0       0   
514              Perso0l Branding                    0       0   
515                  Perso0l Data                    0       0   
516                Perso0l Fi0nce                    0       0   
517                Perso0l Health                    0       0   
518                Perso0lization                    0       0   
534           Predictive A0lytics                    0       0   
546        Professio0l Networking                    0       0   
547          Professio0l Services                    0       0   
548             Project Ma0gement                    0       0   
549                    Promotio0l                    0       0   
550            Property Ma0gement                    0       0   
575                Risk Ma0gement                    0       0   
612        Social Media Ma0gement                    0       0   
633        Supply Chain Ma0gement                    0       0   
637                Task Ma0gement                    0       0   
664                     Veteri0ry                    0       0   
672               Waste Ma0gement                    0       0   

     Cleantech / Semiconductors  Entertainment  Health  Manufacturing  \
20                            0              0       1              0   
22                            0              0       0              0   
59                            0              0       0              0   
79                            0              0       0              0   
89                            0              0       1              0   
91               

Thus we can clearly see that here the word "na" is replaced by "0". In order to use this data properly, we can simply replace the values "0" with "na" but there are some instances where the character "0" is in the name of the category itself.
There is only one such instance in our list which is "Enterprise 2.0". In order to handle this we will first check for the condition where the category name contains "0" and then check for whether the category name ends with the character "0" if it does, we won't replace else we will.

In [56]:
def str_rplc(w):
    if w.startswith('0')==True:
        w=w.replace('0', 'Na')
    elif w.endswith('0')==False:
        w=w.replace('0', 'na')
    return w
mapping.category_list=mapping.category_list.apply(str_rplc)


In [57]:
#preprocessing mapping file
mappingv1=pd.melt(mapping,id_vars='category_list',var_name='main_sector')
mappingv1=mappingv1[mappingv1.value!=0]
mappingv1=mappingv1.drop(['value'],axis=1).reset_index(drop=True)
display(mappingv1.head(5))

category_list          main_sector
0  Adventure Travel  Automotive & Sports
1         Aerospace  Automotive & Sports
2              Auto  Automotive & Sports
3   Automated Kiosk  Automotive & Sports
4        Automotive  Automotive & Sports

In [58]:
#Merging the mapping processed df with the master_frame
master_frame=pd.merge(master_frame,mappingv1,how='left',left_on='primary_sector',right_on='category_list')
display(master_frame.head(5))

company_permalink  \
0             /Organization/-Fame   
1          /Organization/-Qounter   
2  /Organization/004-Technologies   
3            /Organization/0Xdata   
4            /Organization/0Xdata   

                           funding_round_permalink funding_round_type  \
0  /funding-round/9a01d05418af9f794eebff7ace91f638            venture   
1  /funding-round/22dacff496eb7acb2b901dec1dfe5633            venture   
2  /funding-round/1278dd4e6a37fa4b7d7e06c21b3c1830            venture   
3  /funding-round/3bb2ee4a2d89251a10aaa735b1180e44            venture   
4  /funding-round/ae2a174c06517c2394aed45006322a7e            venture   

  funding_round_code   funded_at  raised_amount_usd  \
0                  B  05-01-2015        10000000.00   
1                  A  14-10-2014               0.00   
2                     24-07-2014               0.00   
3                  B  09-11-2015        20000000.00   
4                     03-01-2013         1700000.00   

                        permalink              name  \
0             /Organization/-Fame             #fame   
1          /Organization/-Qounter          :Qounter   
2  /Organization/004-Technologies  004 Technologies   
3            /Organization/0Xdata            H2O.ai   
4            /Organization/0Xdata            H2O.ai   

                        homepage_url  \
0                 http://livfame.com   
1             http://www.qounter.com   
2  http://004gmbh.de/en/004-interact   
3                     http://h2o.ai/   
4                     http://h2o.ai/   

                                     category_list_x     status country_code  \
0                                              Media  operating          IND   
1  Application Platforms|Real Time|Social Network...  operating          USA   
2                                           Software  operating          USA   
3                                          Analytics  operating          USA   
4                                          Analytics  operating          USA   

  state_code                 region           city  founded_at  \
0         16                 Mumbai         Mumbai         NaN   
1         DE             DE - Other  Delaware City  04-09-2014   
2         IL  Springfield, Illinois      Champaign  01-01-2010   
3         CA            SF Bay Area  Mountain View  01-01-2011   
4         CA            SF Bay Area  Mountain View  01-01-2011   

          primary_sector        category_list_y  \
0                  Media                  Media   
1  Application Platforms  Application Platforms   
2               Software               Software   
3              Analytics              Analytics   
4              Analytics              Analytics   

                               main_sector  
0                            Entertainment  
1               News, Search and Messaging  
2                                   Others  
3  Social, Finance, Analytics, Advertising  
4  Social, Finance, Analytics, Advertising

### Sector Analysis 2

Now we will create three different dataframes for the three english speaking countries and the funding as venture.

In [59]:
D1 = master_frame[master_frame.country_code=="USA"]
D1['main_sector_investment_count']=D1.groupby('main_sector')['raised_amount_usd'].transform('count')
D1['main_sector_investment_sum']=D1.groupby('main_sector')['raised_amount_usd'].transform('sum')
D1=D1.reset_index(drop=True)
display(D1.head(5))

c:\users\u21a19\appdata\local\programs\python\python36\lib\site-packages\ipykernel_launcher.py:2: SettingWithCopyWarning:


A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy

c:\users\u21a19\appdata\local\programs\python\python36\lib\site-packages\ipykernel_launcher.py:3: SettingWithCopyWarning:


A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy



company_permalink  \
0          /Organization/-Qounter   
1  /Organization/004-Technologies   
2            /Organization/0Xdata   
3            /Organization/0Xdata   
4            /Organization/0Xdata   

                           funding_round_permalink funding_round_type  \
0  /funding-round/22dacff496eb7acb2b901dec1dfe5633            venture   
1  /funding-round/1278dd4e6a37fa4b7d7e06c21b3c1830            venture   
2  /funding-round/3bb2ee4a2d89251a10aaa735b1180e44            venture   
3  /funding-round/ae2a174c06517c2394aed45006322a7e            venture   
4  /funding-round/e1cfcbe1bdf4c70277c5f29a3482f24e            venture   

  funding_round_code   funded_at  raised_amount_usd  \
0                  A  14-10-2014               0.00   
1                     24-07-2014               0.00   
2                  B  09-11-2015        20000000.00   
3                     03-01-2013         1700000.00   
4                  A  19-07-2014         8900000.00   

                        permalink              name  \
0          /Organization/-Qounter          :Qounter   
1  /Organization/004-Technologies  004 Technologies   
2            /Organization/0Xdata            H2O.ai   
3            /Organization/0Xdata            H2O.ai   
4            /Organization/0Xdata            H2O.ai   

                        homepage_url  \
0             http://www.qounter.com   
1  http://004gmbh.de/en/004-interact   
2                     http://h2o.ai/   
3                     http://h2o.ai/   
4                     http://h2o.ai/   

                                     category_list_x  \
0  Application Platforms|Real Time|Social Network...   
1                                           Software   
2                                          Analytics   
3                                          Analytics   
4                                          Analytics   

              ...             country_code state_code                 region  \
0             ...                      USA         DE             DE - Other   
1             ...                      USA         IL  Springfield, Illinois   
2             ...                      USA         CA            SF Bay Area   
3             ...                      USA         CA            SF Bay Area   
4             ...                      USA         CA            SF Bay Area   

            city  founded_at         primary_sector        category_list_y  \
0  Delaware City  04-09-2014  Application Platforms  Application Platforms   
1      Champaign  01-01-2010               Software               Software   
2  Mountain View  01-01-2011              Analytics              Analytics   
3  Mountain View  01-01-2011              Analytics              Analytics   
4  Mountain View  01-01-2011              Analytics              Analytics   

                               main_sector main_sector_investment_count  \
0               News, Search and Messaging                      4618.00   
1                                   Others                      8767.00   
2  Social, Finance, Analytics, Advertising                      7822.00   
3  Social, Finance, Analytics, Advertising                      7822.00   
4  Social, Finance, Analytics, Advertising                      7822.00   

   main_sector_investment_sum  
0              45600521085.00  
1              82822211972.00  
2              86331402349.00  
3              86331402349.00  
4              86331402349.00  

[5 rows x 21 columns]

In [60]:
D2 = master_frame[master_frame.country_code=="GBR"]
D2['main_sector_investment_count']=D2.groupby('main_sector')['raised_amount_usd'].transform('count')
D2['main_sector_investment_sum']=D2.groupby('main_sector')['raised_amount_usd'].transform('sum')
D2=D2.reset_index(drop=True)
display(D2.head(5))

c:\users\u21a19\appdata\local\programs\python\python36\lib\site-packages\ipykernel_launcher.py:2: SettingWithCopyWarning:


A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy

c:\users\u21a19\appdata\local\programs\python\python36\lib\site-packages\ipykernel_launcher.py:3: SettingWithCopyWarning:


A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy



company_permalink  \
0  /Organization/10-Minutes-With   
1          /Organization/31Dover   
2        /Organization/365Scores   
3        /Organization/365Scores   
4             /Organization/3Sun   

                           funding_round_permalink funding_round_type  \
0  /funding-round/0faccbbcc5818dc5326469f13f5a8ac8            venture   
1  /funding-round/b95cb5a74632e596e19a845e405ef14b            venture   
2  /funding-round/48212f931f542fdef78810bc87aef086            venture   
3  /funding-round/493f78ea0ca33cfac48a57b2351b154b            venture   
4  /funding-round/f1d8c6491b45bcf2a35ef1cf5cae96c4            venture   

  funding_round_code   funded_at  raised_amount_usd  \
0                  A  09-10-2014         4000000.00   
1                  B  01-03-2014         2274716.00   
2                  B  29-09-2014         5500000.00   
3                  A  28-01-2013         1200000.00   
4                     03-03-2014        15152514.00   

                       permalink             name              homepage_url  \
0  /Organization/10-Minutes-With  10 Minutes With  http://10minuteswith.com   
1          /Organization/31Dover          31Dover    http://www.31dover.com   
2        /Organization/365Scores        365Scores  http://biz.365scores.com   
3        /Organization/365Scores        365Scores  http://biz.365scores.com   
4             /Organization/3Sun             3sun      http://3sungroup.com   

                     category_list_x             ...             country_code  \
0                          Education             ...                      GBR   
1        E-Commerce|Wine And Spirits             ...                      GBR   
2  Android|Apps|iPhone|Mobile|Sports             ...                      GBR   
3  Android|Apps|iPhone|Mobile|Sports             ...                      GBR   
4                   Clean Technology             ...                      GBR   

  state_code  region    city  founded_at    primary_sector   category_list_y  \
0         H9  London  London  01-01-2013         Education         Education   
1         H9  London  London  01-07-2012        E-Commerce        E-Commerce   
2         H9  London  London  01-01-2008           Android           Android   
3         H9  London  London  01-01-2008           Android           Android   
4         I9     NaN     NaN  01-01-2007  Clean Technology  Clean Technology   

                               main_sector main_sector_investment_count  \
0                                   Others                       580.00   
1                                   Others                       580.00   
2  Social, Finance, Analytics, Advertising                       481.00   
3  Social, Finance, Analytics, Advertising                       481.00   
4               Cleantech / Semiconductors                       466.00   

   main_sector_investment_sum  
0               4492219646.00  
1               4492219646.00  
2               3773328672.00  
3               3773328672.00  
4               5073290093.00  

[5 rows x 21 columns]

In [61]:
D3 = master_frame[master_frame.country_code=="IND"]
D3['main_sector_investment_count']=D3.groupby('main_sector')['raised_amount_usd'].transform('count')
D3['main_sector_investment_sum']=D3.groupby('main_sector')['raised_amount_usd'].transform('sum')
D3=D3.reset_index(drop=True)
display(D3.head(5))

c:\users\u21a19\appdata\local\programs\python\python36\lib\site-packages\ipykernel_launcher.py:2: SettingWithCopyWarning:


A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy

c:\users\u21a19\appdata\local\programs\python\python36\lib\site-packages\ipykernel_launcher.py:3: SettingWithCopyWarning:


A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy



company_permalink  \
0                 /Organization/-Fame   
1             /Organization/1000Lookz   
2      /Organization/21Diamonds-India   
3  /Organization/247-Learning-Private   
4                 /Organization/3Dsoc   

                           funding_round_permalink funding_round_type  \
0  /funding-round/9a01d05418af9f794eebff7ace91f638            venture   
1  /funding-round/e07f6e580046d00c7aeb33fafb9d90a3            venture   
2  /funding-round/6de7ffef8091ba9f33821f4b861f434a            venture   
3  /funding-round/4cd5cd165bc869bc40705a9e4b626bee            venture   
4  /funding-round/bbedf1410fddeed341018583bbd0ca1e            venture   

  funding_round_code   funded_at  raised_amount_usd  \
0                  B  05-01-2015        10000000.00   
1                     22-07-2013               0.00   
2                  C  15-11-2012         6369507.00   
3                     06-11-2007         4000000.00   
4                  A  01-12-2007         1240000.00   

                            permalink           name  \
0                 /Organization/-Fame          #fame   
1             /Organization/1000Lookz      1000Lookz   
2      /Organization/21Diamonds-India     21Diamonds   
3  /Organization/247-Learning-Private  24x7 Learning   
4                 /Organization/3Dsoc          3DSoC   

                  homepage_url           category_list_x  \
0           http://livfame.com                     Media   
1         http://1000lookz.com                    Beauty   
2     http://www.21diamonds.de                E-Commerce   
3  http://www.24x7learning.com  EdTech|Education|Systems   
4         http://www.3dsoc.com                 3D|Mobile   

              ...             country_code state_code     region       city  \
0             ...                      IND         16     Mumbai     Mumbai   
1             ...                      IND         25    Chennai    Chennai   
2             ...                      IND         10  New Delhi    Gurgaon   
3             ...                      IND         19  Bangalore  Bangalore   
4             ...                      IND         19  Bangalore  Bangalore   

   founded_at primary_sector category_list_y    main_sector  \
0         NaN          Media           Media  Entertainment   
1  01-01-2008         Beauty          Beauty         Health   
2  01-06-2012     E-Commerce      E-Commerce         Others   
3  01-01-2001         EdTech          EdTech         Others   
4  01-06-2006             3D              3D  Manufacturing   

  main_sector_investment_count  main_sector_investment_sum  
0                           90                976191777.00  
1                           56                564370000.00  
2                          332               6329507440.00  
3                          332               6329507440.00  
4                           64                730260000.00  

[5 rows x 21 columns]

Now to we will analyze the three different countries based on the metrics provided in the assignment.

In [62]:
#Function to calculate required metrics for the df
def metrics_calc(df):
    total_investment_count=df.raised_amount_usd.count()
    print('total number of investments : {}'.format(total_investment_count))
    total_investment_amount=df.raised_amount_usd.sum()
    print('total amount of investment : {}'.format(total_investment_amount))
    top_sector=df[['main_sector','main_sector_investment_count']].drop_duplicates().sort_values('main_sector_investment_count',ascending=False).reset_index(drop=True).loc[0][0]
    print('Top sector (based on count of investments) : {}'.format(top_sector))
    second_top_sector=df[['main_sector','main_sector_investment_count']].drop_duplicates().sort_values('main_sector_investment_count',ascending=False).reset_index(drop=True).loc[1][0]
    print('Second-best sector (based on count of investments) : {}'.format(second_top_sector))
    third_top_sector=df[['main_sector','main_sector_investment_count']].drop_duplicates().sort_values('main_sector_investment_count',ascending=False).reset_index(drop=True).loc[2][0]
    print('Third-best sector (based on count of investments) : {}'.format(third_top_sector))
    top_sector_investment_count=df[['main_sector','main_sector_investment_count']].drop_duplicates().sort_values('main_sector_investment_count',ascending=False).reset_index(drop=True).loc[0][1]
    print('Number of investments in the top sector : {}'.format(top_sector_investment_count))
    second_top_sector_investment_count=df[['main_sector','main_sector_investment_count']].drop_duplicates().sort_values('main_sector_investment_count',ascending=False).reset_index(drop=True).loc[1][1]
    print('Number of investments in the second-best sector : {}'.format(second_top_sector_investment_count))
    third_top_sector_investment_count=df[['main_sector','main_sector_investment_count']].drop_duplicates().sort_values('main_sector_investment_count',ascending=False).reset_index(drop=True).loc[2][1]
    print('Number of investments in the third-best sector : {}'.format(third_top_sector_investment_count))
    top_sector_highest_company=df[df.main_sector==top_sector].groupby('company_permalink')[['raised_amount_usd']].sum().sort_values('raised_amount_usd',ascending=False).reset_index().loc[0][0]
    print('Company received the highest investment for the top sector : {}'.format(top_sector_highest_company))
    second_best_sector_highest_company=df[df.main_sector==second_top_sector].groupby('company_permalink')[['raised_amount_usd']].sum().sort_values('raised_amount_usd',ascending=False).reset_index().loc[0][0]
    print('Company received the highest investment for the second based sector : {}'.format(second_best_sector_highest_company))    

In [63]:
metrics_calc(D1)

total number of investments : 38372
total amount of investment : 422510842796.0
Top sector (based on count of investments) : Others
Second-best sector (based on count of investments) : Cleantech / Semiconductors
Third-best sector (based on count of investments) : Social, Finance, Analytics, Advertising
Number of investments in the top sector : 8767.0
Number of investments in the second-best sector : 8268.0
Number of investments in the third-best sector : 7822.0
Company received the highest investment for the top sector : /Organization/Social-Finance
Company received the highest investment for the second based sector : /Organization/Freescale


In [64]:
metrics_calc(D2)

total number of investments : 2303
total amount of investment : 20245627416.0
Top sector (based on count of investments) : Others
Second-best sector (based on count of investments) : Social, Finance, Analytics, Advertising
Third-best sector (based on count of investments) : Cleantech / Semiconductors
Number of investments in the top sector : 580.0
Number of investments in the second-best sector : 481.0
Number of investments in the third-best sector : 466.0
Company received the highest investment for the top sector : /Organization/Oneweb
Company received the highest investment for the second based sector : /Organization/Powa-Technologies


In [65]:
metrics_calc(D3)

total number of investments : 992
total amount of investment : 14391858718.0
Top sector (based on count of investments) : Others
Second-best sector (based on count of investments) : Social, Finance, Analytics, Advertising
Third-best sector (based on count of investments) : News, Search and Messaging
Number of investments in the top sector : 332
Number of investments in the second-best sector : 193
Number of investments in the third-best sector : 154
Company received the highest investment for the top sector : /Organization/Flipkart
Company received the highest investment for the second based sector : /Organization/Shopclues-Com


## Plots

#### Plot showing the fraction of total investments (globally) in venture, seed, and private equity, and the average amount of investment in each funding type :

In [72]:
import plotly.express as px
import plotly.graph_objects as go
data = rounds2

data.raised_amount_usd=pd.to_numeric(data.raised_amount_usd)
grouped=data.groupby('funding_round_type')['raised_amount_usd'].agg(['mean','sum']).reset_index()
grouped=grouped[grouped.funding_round_type.isin(['angel','venture','seed','private_equity'])].reset_index(drop=True)
total_investment=rounds2.raised_amount_usd.sum()
grouped['fraction']=grouped['sum']/total_investment

print(grouped)
print(total_investment)

fig = px.bar(grouped, x='funding_round_type', y='mean',
             hover_data=['fraction'], color='fraction',
             labels={'funding_round_type':'Funding round type','mean':'average amount of investment'}, height=400)
fig.add_shape(
        go.layout.Shape(
            type="line",
            y0=5000000,
            y1=5000000,
        ))
fig.add_shape(
        go.layout.Shape(
            type="line",
            y0=15000000,
            y1=15000000,
        ))
fig.update_layout(title_text="Fraction of total investments (globally) in venture, seed, and private equity")
fig.show()

  funding_round_type        mean             sum  fraction
0              angel   764564.35   4659255123.00      0.00
1     private_equity 62111788.23 141925436105.00      0.14
2               seed   556606.74  16989864181.00      0.02
3            venture 10634054.44 590126216876.00      0.60
990125085135.0


As we can see the only funding round type which has the investment between 5M to 15M is venture and thats the reason it is most suitable for Spark Funds.

* Plot showing the top 9 countries against the total amount of investments of funding type FT. 

In [73]:
#For this we have the dataframe top9, we can use that here.
datav1=top9.reset_index()
#As we know the top three countries, so we will add a new column color to distinguish them.
datav1['color']=np.where(datav1['country_code'].isin(['USA','GBR','IND']), 'green', 'red')
display(datav1)

country_code  raised_amount_usd  color
0          USA    422510842796.00  green
1          CHN     39835418773.00    red
2          GBR     20245627416.00  green
3          IND     14391858718.00  green
4          CAN      9583332317.00    red
5          FRA      7259536732.00    red
6          ISR      6907514579.00    red
7          DEU      6346959822.00    red
8          JPN      3363676611.00    red

In [75]:
fig = px.bar(datav1, x='country_code', y='raised_amount_usd',
             hover_data=['color'], color='color',
             labels={'country_code':'Country','raised_amount_usd':'total amount of investments for venture'}, height=400)
fig.update_layout(showlegend=False)
fig.update_layout(title_text="Top 9 countries against the total amount of investments of funding type FT")
fig.show()


* Plot showing the number of investments in the top 3 sectors of the top 3 countries

In [69]:
#For this we can use the previous D1, D2 and D3 dataframes.
#Getting the top 3 sectors information from a df 
D1_top_3=pd.DataFrame(D1.groupby(['country_code','main_sector'])[['main_sector_investment_count']].mean().sort_values(by='main_sector_investment_count',ascending=False).reset_index().head(3))
D2_top_3=pd.DataFrame(D2.groupby(['country_code','main_sector'])[['main_sector_investment_count']].mean().sort_values(by='main_sector_investment_count',ascending=False).reset_index().head(3))
D3_top_3=pd.DataFrame(D3.groupby(['country_code','main_sector'])[['main_sector_investment_count']].mean().sort_values(by='main_sector_investment_count',ascending=False).reset_index().head(3))
combined =pd.concat([D1_top_3,D2_top_3,D3_top_3])
display(combined)

country_code                              main_sector  \
0          USA                                   Others   
1          USA               Cleantech / Semiconductors   
2          USA  Social, Finance, Analytics, Advertising   
0          GBR                                   Others   
1          GBR  Social, Finance, Analytics, Advertising   
2          GBR               Cleantech / Semiconductors   
0          IND                                   Others   
1          IND  Social, Finance, Analytics, Advertising   
2          IND               News, Search and Messaging   

   main_sector_investment_count  
0                       8767.00  
1                       8268.00  
2                       7822.00  
0                        580.00  
1                        481.00  
2                        466.00  
0                        332.00  
1                        193.00  
2                        154.00

In [77]:
fig = px.bar(combined, x="country_code", y="main_sector_investment_count",labels={'main_sector_investment_count':'Investment Count','country_code':'Country'}, color='main_sector', barmode='group',
             height=400).for_each_trace(lambda t: t.update(name=t.name.replace("main_sector=","")))
fig.update_layout(title_text="Number of investments in the top 3 sectors of the top 3 countries")
fig.show()